[TW] Pet fix (hit, haste, feral spirit base stats)

Diryn90 opened the pull request:
- Hunter pet receive hit based on hunter ranged hit.
- Other pet/guardian receive hit based on the owner melee hit.
- Ghoul pet receive owner haste.
- Feral spirit have the correct base stats (from script)

Nyeriah commented on src/server/game/Entities/Pet/Pet.cpp:
```diff
@@ -1069,6 +1070,19 @@ bool Guardian::InitStatsForLevel(uint8 petlevel)
+        Player* player = GetOwner()->ToPlayer();
+        if (IsHunterPet())
+            m_modMeleeHitChance = player->GetRatingBonusValue(CR_HIT_RANGED);
+        else
```
Nope this is not correct, not every pet inherit stats in the same way. For example gorgoyles should get the dk's melee hit rating as spell hit rating, along other things.

Unholychick replied:
worth mentioning is that you must differ between "pets" and "guardians", the gorgoyle is a guardian, not a pet. water elementals, spirit wolves, etc are all guardians

Diryn90 commented:
At the moment the only haste value that I fixed is for the ghoul.

See the method that I added in StatSystem.cpp

I didn't have a look yet for gargoyle.

Diryn90 commented:
For hit rating, all hunter pet should get the range rating of the hunter, i read that on some forums. And for other pets, if they actually use melee the hit value should come from the melee owner hit value. For pets like imp, they don't even do melee so it will be without effect.
I can move the hit part for the wolves only but i thought it would be more correct to apply this rule for other pets.

Unholychick commented on src/server/game/Entities/Unit/StatSystem.cpp:
```diff
@@ -806,13 +806,26 @@ void Player::UpdateArmorPenetration(int32 amount)
+    m_modMeleeHitChance += crHit;
+
+    Pet* pet = GetPet();                               
+    Guardian* guardian = GetGuardianPet();
```
```if (Guardian* guardian = GetGuardianPet())```

Diryn90 commented:
I have done another commit to limit the gain of the owner's hit only for Pet and for Spirit wolf.

Nyeriah commented on src/server/game/Entities/Pet/Pet.cpp:
```diff
@@ -1069,6 +1070,18 @@ bool Guardian::InitStatsForLevel(uint8 petlevel)
+        Player* player = GetOwner()->ToPlayer();
+        if (IsHunterPet())
+            m_modMeleeHitChance = player->GetRatingBonusValue(CR_HIT_RANGED);
+        else if (IsPet() || IsSpiritWolf())
```
Did you mean && instead of || ?

As it it's still applied to every pet

Diryn90 replied:
It should apply to any pet.

The OR is just to also apply to spirit wolf which does not inherit Pet class but still need this rule applied.

For the hunter this rule also apply but uses the hunter's range hit instead of melee hit.
